feat: Implemented event invoice table for organizers

mrsaicharan1 opened the pull request:
<!-- 
(Thanks for sending a pull request! Please make sure you click the link above to view the contribution guidelines, then fill out the blanks below.)
-->

<!-- Add the issue number that is fixed by this PR (In the form Fixes #123) -->

Fixes #3118 

#### Short description of what this resolves:
This feature adds the tables required for event invoices which shows the upcoming, paid and due event invoices along with relevant information. 
#### Screenshots
As soon as the server implementation for event invoices are functional, the event invoices will show up here along with their respective invoice links.
<img width="1244" alt="Screenshot 2019-07-31 at 3 06 51 PM" src="https://user-images.githubusercontent.com/25197147/62201367-e3d66d80-b3a4-11e9-8a6e-cbe94b6a5bce.png">



#### Changes proposed in this pull request:

- Filter upcoming, paid and due invoices
- Added controller for event-invoices
- Designed template for event invoices

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia).
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The acceptance, integration, unit tests and linter pass locally with my changes <!-- use `ember test` to run all the tests -->
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

kushthedude commented:
@mrsaicharan1 Please put it on-hold, I dont think you would like to refactor it twice , Ember-Tables are almost ready

Anupam-dagar commented on tests/unit/controllers/account/billing-info/invoices-test.js:
```diff
@@ -0,0 +1,12 @@
+module('Unit | Controller | account/billing-info/invoices', function(hooks) {
+  setupTest(hooks);
+
+  // Replace this with your real tests.
```
This comment is not really needed

mrsaicharan1 replied:
Yeah, will remove this test. It got generated automatically when I used the ember CLI to generate the controller.

Anupam-dagar commented on tests/unit/controllers/account/billing-info/invoices-test.js:
```diff
@@ -0,0 +1,12 @@
+import { module, test } from 'qunit';
+import { setupTest } from 'ember-qunit';
+
+module('Unit | Controller | account/billing-info/invoices', function(hooks) {
```
this test is failing in travis

mrsaicharan1 replied:
@Anupam-dagar Some other test seems to be failing :(

niranjan94 commented on app/controllers/account/billing-info/invoices.js:
```diff
@@ -0,0 +1,34 @@
+export default Controller.extend({
+  dueInvoices      : filterBy('model', 'status', 'due'),
+  paidInvoices     : filterBy('model', 'status', 'paid'),
+  upcomingInvoices : computed(function() {
```
A computed property _must_ not return a promise. This should happen at the route level.

mrsaicharan1 replied:
@niranjan94 Moved it to the routes and returned a hash of promises

niranjan94 commented on app/routes/account/billing-info/invoices.js:
```diff
@@ -1,4 +1,34 @@
 
 export default Route.extend({
+  async model() {
+    let filterOptions = [];
```
this declaration of empty array is unnecessary ..

mrsaicharan1 replied:
Travis seems to fail with this message when I remove it.. .`ReferenceError: filterOptions is not defined`

niranjan94 replied:
yes. you should be doing 

```js
const filterOptions = [
      {
        and: [
          {
            name : 'deleted-at',
            op   : 'eq',
            val  : null
          },
          {
            name : 'created-at',
            op   : 'ge',
            val  : moment().subtract(30, 'days').toISOString()
          }
        ]
      }
    ]
```

niranjan94 commented on app/routes/account/billing-info/invoices.js:
```diff
@@ -1,4 +1,34 @@
+        ]
+      }
+    ];
+    return RSVP.hash({
```
Use async-await instead.

mrsaicharan1 commented:
> Use ember tables?

Yeah, just saw your merged PR. I  considered refactoring after this was merged but I'll do it right now as the ember tables are functional

niranjan94 commented on app/controllers/account/billing-info/invoices.js:
```diff
@@ -0,0 +1,8 @@
+import Controller from '@ember/controller';
+import { filterBy } from '@ember/object/computed';
+
+export default Controller.extend({
```
Lets start moving to ES6 style classes for all new resources. 
https://github.com/fossasia/open-event-frontend/pull/3169/files

niranjan94 commented on app/routes/account/billing-info/invoices.js:
```diff
@@ -1,4 +1,33 @@
+import RSVP from 'rsvp';
+import moment from 'moment';
 
 export default Route.extend({
```
Lets start moving to ES6 style classes for all new resources. 
https://github.com/fossasia/open-event-frontend/pull/3169/files

kushthedude commented:
Use ember tables for all pending and past too.

mrsaicharan1 commented:
> Use ember tables for all pending and past too.

Pending & past invoices ? What does that mean?

niranjan94 commented on app/templates/account/billing-info/invoices.hbs:
```diff
@@ -1,14 +1,115 @@
+                  fillMode="equal-column"
+    }}
   </div>
+    {{!-- <table class="ui stackable structured celled compact table">
```
Remove the commented parts if not required

niranjan94 commented on app/controllers/account/billing-info/invoices.js:
```diff
@@ -0,0 +1,42 @@
+
+export default class extends Controller {
+  queryParams = ['page', 'per_page'];
+  page = 1;
```
I think @CosmicCoder96 has added mixins for this boilerplate code. Use that

niranjan94 commented on app/controllers/account/billing-info/invoices.js:
```diff
@@ -0,0 +1,42 @@
+  search = null;
+  sort_dir = null;
+  sort_by = null;
+  dueInvoices  = filterBy('model.eventInvoices', 'status', 'due');
```
Should be a decorator

niranjan94 commented on app/controllers/account/billing-info/invoices.js:
```diff
@@ -0,0 +1,42 @@
+  sort_dir = null;
+  sort_by = null;
+  dueInvoices  = filterBy('model.eventInvoices', 'status', 'due');
+  paidInvoices = filterBy('model.eventInvoices', 'status', 'paid');
```
Should be a decorator

kushthedude commented:
> > Use ember tables for all pending and past too.
> 
> Pending & past invoices ? What does that mean?

You understood it correctly 😄

kushthedude commented on tests/unit/routes/app/templates/account/billing-info/invoices/list-test.js:
```diff
@@ -0,0 +1,11 @@
+  setupTest(hooks);
+
+  test('it exists', function(assert) {
+    let route = this.owner.lookup('route:app/templates/account/billing-info/invoices/list');
```
It would be great if you can you write render test and Table test as done in others components rather than route test .

niranjan94 commented on app/controllers/account/billing-info/invoices.js:
```diff
@@ -0,0 +1,91 @@
+import EmberTableMixin from 'open-event-frontend/mixins/ember-table';
+	
+export default class extends Controller.extend(EmberTableMixin) {
+  // queryParams = ['page', 'per_page'];
```
Remove if not required

niranjan94 commented on app/controllers/account/billing-info/invoices.js:
```diff
@@ -0,0 +1,91 @@
+  }
+
+  @computed('dueInvoices')
+  get dueInvoiceRows() {
```
Why is this needed ? Can't `dueInvoices` be used directly ?

mrsaicharan1 replied:
Did you mean that `dueInvoices` can be used directly instead of watching for them?

niranjan94 replied:
Yes.

niranjan94 commented on app/controllers/account/billing-info/invoices.js:
```diff
@@ -0,0 +1,91 @@
+  }
+
+  @computed('paidInvoices')
+  get paidInvoiceRows() {
```
Use `paidInvoices` directly

niranjan94 commented on app/controllers/account/billing-info/invoices.js:
```diff
@@ -0,0 +1,91 @@
+  }
+
+  @computed('upcomingInvoices')
+  get upcomingInvoiceRows() {
```
Use `upcomingInvoices` directly

mrsaicharan1 commented:
@niranjan94 @CosmicCoder96 @uds5501 @shreyanshdwivedi I've integrated ember tables for the event invoice section. This PR is now ready for a review.

niranjan94 commented on app/controllers/account/billing-info/invoices/list.js:
```diff
@@ -0,0 +1,193 @@
+export default class extends Controller.extend(EmberTableControllerMixin) {
+
+  @computed()
+  get upcomingInvoiceColumns() {
```
These are arrays with plain objects. Can be a normal property instead of a computed property

mrsaicharan1 replied:
@niranjan94 Replaced these with POJOs.

mrsaicharan1 replied:
@niranjan94 Replaced these with POJOs. Please have a look.

niranjan94 commented on app/controllers/account/billing-info/invoices/list.js:
```diff
@@ -0,0 +1,193 @@
+  }
+
+  @computed()
+  get AllColumns() {
```
These are arrays with plain objects. Can be a normal property instead of a computed property

niranjan94 commented on app/controllers/account/billing-info/invoices/list.js:
```diff
@@ -0,0 +1,193 @@
+  }
+
+  @computed()
+  get paidInvoiceColumns() {
```
These are arrays with plain objects. Can be a normal property instead of a computed property

niranjan94 commented on app/controllers/account/billing-info/invoices/list.js:
```diff
@@ -0,0 +1,193 @@
+  }
+
+  @computed()
+  get dueInvoiceColumns() {
```
These are arrays with plain objects. Can be a normal property instead of a computed property

niranjan94 commented on app/controllers/account/billing-info/invoices/list.js:
```diff
@@ -0,0 +1,193 @@
+  }
+
+  @computed()
+  get allInvoiceColumns() {
```
These are arrays with plain objects. Can be a normal property instead of a computed property

niranjan94 commented on app/controllers/account/billing-info/invoices/list.js:
```diff
@@ -0,0 +1,161 @@
+import { computed } from '@ember/object';
+import EmberTableControllerMixin from 'open-event-frontend/mixins/ember-table-controller';
+
+export default class extends Controller.extend(EmberTableControllerMixin) {
```
Check alignment on this file. Seems misaligned.
Also, use the model directly instead of an unnecessary computed property in between

shreyanshdwivedi commented:
@niranjan94 I'm following changes asked by you on this PR closely and I'll be making the changes to my PR of organizer session table. Will tag you once ready for review again

shreyanshdwivedi commented on app/controllers/account/billing-info/invoices/list.js:
```diff
@@ -0,0 +1,156 @@
+
+get dueInvoiceRows() {
+  const rows = [];
+  this.model.data.forEach(row => {
```
I think we can use a `this.model.data.map` here and for other such iterations too.

mrsaicharan1 commented:
Had to change the route billing-info -> billing to comply with the recent change. Made the requested changes to this. Please have a look. 
Will proceed with the payment workflow as soon as this gets merged.

niranjan94 commented on app/controllers/account/billing/invoices/list.js:
```diff
@@ -0,0 +1,156 @@
+import Controller from '@ember/controller';
+import EmberTableControllerMixin from 'open-event-frontend/mixins/ember-table-controller';
+
+export default class extends Controller.extend(EmberTableControllerMixin) {
```
@mrsaicharan1 the spacing is incorrect in this file. Please fix it.

niranjan94 commented on app/controllers/account/billing/invoices/list.js:
```diff
@@ -0,0 +1,156 @@
+
+get paidInvoiceRows() {
+  const rows = [];
+  this.model.data.map(row => {
```
Use model directly

mrsaicharan1 replied:
Do you mean `this.model.data.map(row => {...`? It returns undefined when I do that.

CosmicCoder96 replied:
No, check the other PRs.

niranjan94 commented on app/controllers/account/billing/invoices/list.js:
```diff
@@ -0,0 +1,156 @@
+
+get upcomingInvoiceRows() {
+  const rows = [];
+  this.model.data.map(row => {
```
Use model directly

niranjan94 commented on app/controllers/account/billing/invoices/list.js:
```diff
@@ -0,0 +1,156 @@
+
+get allInvoiceRows() {
+  const rows = [];
+  this.model.data.map(row => {
```
Use model directly

mrsaicharan1 commented:
@CosmicCoder96 Please review this. Would want to proceed with the other parts of this asap

CosmicCoder96 commented on app/controllers/account/billing/invoices/list.js:
```diff
@@ -0,0 +1,155 @@
+  }
+
+];
+allInvoiceColumns =  [
```
make them computed properties.

mrsaicharan1 replied:
@CosmicCoder96 @niranjan94 Had mentioned that computed properties won't be required here. POJOs are enough. [In one of his reviews]

mrsaicharan1 commented:
@CosmicCoder96 I could've used the model directly by passing in `rows=model.data` if the columns were similar for 3 of the categories. But here, for due, upcoming & paid invoices, the row entries which are being pushed are different.

uds5501 commented on app/models/event-invoice.js:
```diff
@@ -26,6 +26,8 @@ export default ModelBase.extend({
   amount        : attr('number'),
   completedAt   : attr('moment'),
+  invoicePdfUrl : attr('string'),
+  eventId       : attr('number'),
```
Do check if it can be achieved using `event.get('originalId')` or something similar

CosmicCoder96 replied:
this is the actual event invoice model, either the server is returning the eventId or not, if it is, then it's definition here is not optional.

mrsaicharan1 commented:
Yeah this is not required anyways ... I'll remove it

On Tue, 30 Jul 2019 at 5:45 PM, Uddeshya Singh <notifications@github.com>
wrote:

> *@uds5501* commented on this pull request.
> ------------------------------
>
> In app/models/event-invoice.js
> <https://github.com/fossasia/open-event-frontend/pull/3230#discussion_r308685753>
> :
>
> > @@ -26,6 +26,8 @@ export default ModelBase.extend({
>    expYear      	: attr('number'),
>    amount        : attr('number'),
>    completedAt   : attr('moment'),
> +  invoicePdfUrl : attr('string'),
> +  eventId       : attr('number'),
>
> Do check if it can be achieved using event.get('originalId') or something
> similar
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/fossasia/open-event-frontend/pull/3230?email_source=notifications&email_token=AGAHUW7LYIQYP2NSXOCULHTQCAWEXA5CNFSM4H5ETMZKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB77KLFY#pullrequestreview-268346775>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGAHUW2J4XRGXNWBOVHVISLQCAWEXANCNFSM4H5ETMZA>
> .
>

mrsaicharan1 commented:
@uds5501 @shreyanshdwivedi @CosmicCoder96 I was able to obtain the event logo & name by creating a cell component & passing in the record. Is this the right way?

CosmicCoder96 commented on app/routes/account/billing/invoices/list.js:
```diff
@@ -0,0 +1,73 @@
+
+    filterOptions = this.applySearchFilters(filterOptions, params, searchField);
+
+    let queryObject = {
```
don't break the pattern, every other table has the name `queryString'

CosmicCoder96 commented on app/routes/account/billing/invoices/list.js:
```diff
@@ -0,0 +1,73 @@
+      include        : 'event',
+      filter         : filterOptions,
+      'page[size]'   : params.per_page || 10,
+      'page[number]' : params.page || 4
```
Why would you start with page 4?

CosmicCoder96 commented:
> I was able to obtain the event logo & name by creating a cell componen

Does it actually work, did you test it?

mrsaicharan1 commented:
> > I was able to obtain the event logo & name by creating a cell componen
> 
> Does it actually work, did you test it?

Yes, I've updated the screenshot too. Please have a look. Also, the `eventId` field was not required & therefore I removed it.

mrsaicharan1 commented:
@pradeepgangwar @uds5501 @shreyanshdwivedi This is ready for another review.

CosmicCoder96 commented on app/templates/account/billing/invoices/list.hbs:
```diff
@@ -0,0 +1,65 @@
+{{#if (eq model.params.invoice_status 'due') }}
```
Too much repitition.

mrsaicharan1 commented:
I have shifted the logic to controllers & removed repetitive template code. Please have a look now.

mrsaicharan1 commented:
@CosmicCoder96 Made the requested changes. Please have a look.
